File: MiniDOMSIP/Dialers/3CX/3CXDialer.cs
using System.IO;
using DMD.XML;

namespace minidom.PBX
{

    /// <summary>
    /// Dialler per usare il software client 3CX
    /// </summary>
    public class C3CXDialer 
        : DialerBaseClass
    {
        private string m_Path;

        /// <summary>
        /// Costruttore
        /// </summary>
        public C3CXDialer()
        {
            m_Path = Get3CXPath();
        }

        /// <summary>
        /// Prepara il numero da comporre
        /// </summary>
        /// <param name="number"></param>
        /// <returns></returns>
        public string PrepareNumber(string number)
        {
            return DMD.Strings.Trim(number);
        }

        /// <summary>
        /// Compone il numero
        /// </summary>
        /// <param name="number"></param>
        public override void Dial(string number)
        {
            if (!IsInstalled())
                return;
            DMD.IOUtils.Shell(m_Path + " dial:" + PrepareNumber(number));
        }

        /// <summary>
        /// Restituisce true se il programma é installato
        /// </summary>
        /// <returns></returns>
        public override bool IsInstalled()
        {
            return !string.IsNullOrEmpty(m_Path);
        }

        private static string Get3CXPath()
        {
            string p = Path.Combine(GetProgramFilesFolder(), @"3CXPhone\3CXPhone.exe");
            if (File.Exists(p))
                return p;
            p = Path.Combine(GetProgramFilesFolder() + " (x86)", @"3CXPhone\3CXPhone.exe");
            if (File.Exists(p))
                return p;
            return DMD.Strings.vbNullString;
        }

        /// <summary>
        /// Nome univoco del dialler
        /// </summary>
        public override string Name
        {
            get
            {
                return "3CX VoIP Phone";
            }
        }

        /// <summary>
        /// Aggancia
        /// </summary>
        public override void HangUp()
        {
        }

        /// <summary>
        /// Restituisce true se i due oggetti sono uguali
        /// </summary>
        /// <param name="obj"></param>
        /// <returns></returns>
        public override bool Equals(object obj)
        {
            if (!(obj is C3CXDialer))
                return false;
            return base.Equals(obj) && (((C3CXDialer)obj).m_Path ?? "") == (m_Path ?? "");
        }
         
    }
}

File: MiniDOMSIP/Dialers/3CXDialer.vb
Imports System.IO

Public Class C3CXDialer
    Inherits DialerBaseClass

    Private m_Path As String

    Public Sub New()
        Me.m_Path = Get3CXPath()
    End Sub

    Function PrepareNumber(ByVal number As String) As String
        Return Trim(number)
    End Function

    Public Overrides Sub Dial(number As String)
        If Not Me.IsInstalled Then Return
        Shell(Me.m_Path & " dial:" & Me.PrepareNumber(number))
    End Sub

    Public Overrides Function IsInstalled() As Boolean
        Return (Me.m_Path <> vbNullString)
    End Function

    Private Shared Function Get3CXPath() As String
        Dim p As String = Path.Combine(GetProgramFilesFolder, "3CXPhone\3CXPhone.exe")
        If (File.Exists(p)) Then Return p
        p = Path.Combine(GetProgramFilesFolder() & " (x86)", "3CXPhone\3CXPhone.exe")
        If (File.Exists(p)) Then Return p
        Return vbNullString
    End Function

    Public Overrides ReadOnly Property Name As String
        Get
            Return "3CX VoIP Phone"
        End Get
    End Property

    Public Overrides Sub HangUp()

    End Sub

    Public Overrides Function Equals(obj As Object) As Boolean
        If Not (TypeOf (obj) Is C3CXDialer) Then Return False
        Return MyBase.Equals(obj) AndAlso DirectCast(obj, C3CXDialer).m_Path = Me.m_Path
    End Function

End Class
